core/domain/story_fetchers.py

```diff
@@ -0,0 +1,325 @@
+    """
+    progress_model = user_models.StoryProgressModel.get(
+        user_id, story_id, strict=False)
+
+    if not progress_model:
+        return []
+
+    num_of_nodes = min(len(progress_model.completed_node_ids), 3)
+    story = get_story_by_id(story_id)
+    ordered_node_ids = (
+        [node.id for node in story.story_contents.get_ordered_nodes()])
+    ordered_completed_node_ids = (
+        [node_id for node_id in ordered_node_ids
+         if node_id in progress_model.completed_node_ids]
+    )
+    return ordered_completed_node_ids[-num_of_nodes:]
+
+
+def get_completed_nodes_in_story(user_id, story_id):
+    """Returns nodes that are completed in a story
+
+    Args:
+        user_id: str. The user id of the user.
+        story_id: str. The id of the story.
+
+    Returns:
+        list(StoryNode): The list of the story nodes that the user has
+        completed.
+    """
+    story = get_story_by_id(story_id)
+    completed_nodes = []
+
+    completed_node_ids = get_completed_node_ids(user_id, story_id)
+    for node in story.story_contents.nodes:
+        if node.id in completed_node_ids:
+            completed_nodes.append(node)
+
+    return completed_nodes
+
+
+def get_pending_nodes_in_story(user_id, story_id):
+    """Returns the nodes that are pending in a story
+
+    Args:
+        user_id: str. The user id of the user.
+        story_id: str. The id of the story.
+
+    Returns:
+        list(StoryNode): The list of story nodes, pending
+        for the user.
+    """
+    story = get_story_by_id(story_id)
+    pending_nodes = []
+
+    completed_node_ids = get_completed_node_ids(user_id, story_id)
+    for node in story.story_contents.nodes:
+        if node.id not in completed_node_ids:
+            pending_nodes.append(node)
+
+    return pending_nodes
+
+
+def get_completed_node_ids(user_id, story_id):
+    """Returns the ids of the nodes completed in the story.
+
+    Args:
+        user_id: str. ID of the given user.
+        story_id: str. ID of the story.
+
+    Returns:
+        list(str). List of the node ids completed in story.
+    """
+    progress_model = user_models.StoryProgressModel.get(
+        user_id, story_id, strict=False)
+
+    return progress_model.completed_node_ids if progress_model else []
+
+
+def get_node_index_by_story_id_and_node_id(story_id, node_id):
+    """Returns the index of the story node with the given story id
+    and node id.
+
+    Args:
+        story_id: str. ID of the story.
+        node_id: str. ID of the story node.
+
+    Returns:
+        int. The index of the corresponding node.
+
+    Raises:
+        Exception. The given story does not exist.
+        Exception. The given node does not exist in the story.
+    """
+    try:
+        story = get_story_by_id(story_id)
+    except Exception:
+        raise Exception('Story with id %s does not exist.' % story_id)
+
+    node_index = story.story_contents.get_node_index(node_id)
+    if node_index is None:
```

Comment:
This part does not seem to be tested; make sure your tests are comprehensive.

Reply:
That was an already existing function and it seems the test for that is present in story_services_test.

Reply:
Probably -- in general, place the tests in the test files which correspond to their code files, so that it's easy to identify which tests go with which code.

Reply:
Done